RHDEVDOCS-3279 Integrate Shared Resources Operator with Cluster Storage (#41898)
openshift/openshift-docs: modules/ephemeral-storage-additional-details-on-volumeattributes-on-shared-resource-pod-volumes.adoc

:_content-type: REFERENCE

[id="ephemeral-storage-additional-details-on-volumeattributes-on-shared-resource-pod-volumes_{context}"]

> = Additional details on VolumeAttributes on shared resource pod volumes

Author: @gabemontero @prietyc123 @bburt-rh New topic. Please review and approve.

> = Additional details on VolumeAttributes on shared resource pod volumes

Author: Red Hat documentation uses sentence-style capitalization instead of the traditional title-style capitalization.

[role="_abstract"]

> The following attributes determine affect shared resource pod volumes in various ways:

Author: Updated. Thanks.

* The `refreshResource` attribute in the `volumeAttributes` properties.
* The `refreshResources` attribute in the Shared Resource CSI Driver configuration.
* The `sharedSecret` and `sharedConfigMap` attributes in the `volumeAttributes` properties.

== The `refreshResource` attribute

The Shared Resource CSI Driver honors the `refreshResource` attribute in `volumeAttributes` properties of the volume. This attribute controls whether updates to the contents of the underlying `Secret` or `ConfigMap` object are copied to the volume *after* the volume is initially provisioned as part of pod startup. The default value of `refreshResource` is `true`, which means that contents are updated.

[IMPORTANT]
====

> The following `refreshResources` attribute overrides the `refreshResource` attribute in `volumeAttributes` properties of the volume.

Reviewer: ```suggestion
If the Shared Resource CSI Driver configuration had disabled the refreshing of Shared Resources (both SharedSecrets and SharedConfigMaps), then the `refreshResource` attribute in the `volumeAttribute` properties has no effect.  The intent of this attribute is to disable refresh for specific Volume mounts.
```

Author: Updated. Thanks.

> If the Shared Resource CSI Driver configuration had disabled the refreshing of both the shared `SharedSecret` and `SharedConfigMap` custom resource (CR) instances, then the `refreshResource` attribute in the `volumeAttribute` properties has no effect. The intent of this attribute is to disable refresh for specific volume mounts.

Reviewer: ```suggestion
If the Shared Resource CSI Driver configuration had disabled the refreshing of both the shared `SharedSecret` and `SharedConfigMap` custom resource (CR) instances, then the `refreshResource` attribute in the `volumeAttribute` properties has no effect. The intent of this attribute is to disable refresh for specific volume mounts when refresh is generally allowed.
```

====

== The `refreshResources` attribute

You can use a global switch to enable or disable refreshing of shared resources. This switch is the `refreshResources` attribute in the `csi-driver-shared-resource-config` config map for the Shared Resource CSI Driver, which you can find in the `openshift-cluster-csi-drivers` namespace. If you set this `refreshResources` attribute to `false`, none of the `Secret` or `ConfigMap`-related content stored in the volume is updated after the initial provisioning of the volume.

[IMPORTANT]
====

> This `refreshResources` attribute overrides the `refreshResource` attribute in `volumeAttributes` properties of the volume.

Reviewer: ```suggestion
Disabling refresh via this configuration on the Shared Resource CSI Driver disable refresh for all Shared Resource CSI Volume mounts in the cluster, regardless of use of the `refreshResources` attribute in a specific CSI Volume's `volumeAttribute` properties.
```

Author: Updated. Thanks. Throughout these topics, I made sure to specify:
- `refreshResource` (SINGULAR) for the attribute in the `volumeAttributes`. 
- `refreshResources` (PLURAL) for the attribute in the `csi-driver-shared-resource-config` config map

====

> 

Author: The current version has distinct admonitions that each describe the two parameters, `refreshResources` and `refreshResource`. Due to their apparent similarity but contrary effects, we feel that using admonitions is justified.

> == Validation of volumeAttributes before provisioning a shared resource volume for a pod

Reviewer: ```suggestion
== Validation of volumeAttributes before provisioning a Shared Resource Volume for a pod
```

Author: I'm applying Red Hat's documentation standard "sentence-case" capitalization for titles. The terms, "shared resources" and "shared resource volumes," are routinely lower-case throughout the OpenShift documentation.

> In the `volumeAttributes` of a single volume, you must set either `sharedSecret` or `sharedConfigMap` attribute to the value of a `SharedSecret` or `SharedConfigMap` resource. Otherwise, during provisioning the volume during pod startup, a validation returns an error to the kubelet.

Reviewer: ```suggestion
In the `volumeAttributes` of a single volume, you must set either `sharedSecret` or `sharedConfigMap` attribute to the value of a `SharedSecret` or `SharedConfigMap` instance. Otherwise, during provisioning the volume during pod startup, the Shared Resource CSI driver returns an error to the kubelet.
```

Author: Updated. Thanks.

> In the `volumeAttributes` of a single volume, you must set either `sharedSecret` or `sharedConfigMap` attribute to the value of a `SharedSecret` or `SharedConfigMap` resource. Otherwise, during provisioning the volume during pod startup, a validation returns an error to the kubelet.

Author: Thank you! Updated.
